samples/siwx91x_ota/src/main.c

```diff
@@ -0,0 +1,579 @@
+		ret = k_event_wait_safe(&ctx->events, NET_IPV4_EVENT_READY, false, K_SECONDS(30));
+	} else {
+		ret = k_event_wait_safe(&ctx->events, NET_IPV6_EVENT_READY, false, K_SECONDS(30));
+	}
+	if (!ret && IS_ENABLED(CONFIG_NET_IPV6)) {
+		ret = k_event_wait_safe(&ctx->events, NET_IPV6_EVENT_READY, false, K_NO_WAIT);
+	}
+	return (ret > 0) ? 0 : -ETIMEDOUT;
+}
+
+/*
+ * @brief Connect to OTA server
+ */
+static int ota_connect_to_server(struct app_ctx *ctx)
+{
+	int peer_verify = TLS_PEER_VERIFY_NONE;
+	struct zsock_addrinfo *res;
+	int ret;
+
+	ret = zsock_getaddrinfo(ctx->url.host, ctx->url.port, NULL, &res);
+	if (ret != 0) {
+		printf("Cannot resolve %s:%s: %s\n", ctx->url.host, ctx->url.port,
+		       zsock_gai_strerror(ret));
+		return ret;
+	}
+
+	if (ctx->url.use_tls) {
+		ctx->sock = zsock_socket(res->ai_family, SOCK_STREAM, IPPROTO_TLS_1_2);
+		if (ctx->sock < 0) {
+			goto error;
+		}
+		ret = zsock_setsockopt(ctx->sock, SOL_TLS, TLS_PEER_VERIFY, &peer_verify,
+				       sizeof(peer_verify));
+		if (ret < 0) {
+			goto error;
+		}
+	} else {
+		/* Use regular TCP for HTTP */
+		ctx->sock = zsock_socket(res->ai_family, SOCK_STREAM, IPPROTO_TCP);
+		if (ctx->sock < 0) {
+			goto error;
+		}
+	}
+
+	/* Connect to the server */
+	ret = zsock_connect(ctx->sock, res->ai_addr, res->ai_addrlen);
+	if (ret < 0) {
+		printf("Connection failed: %s\n", strerror(errno));
+		goto error;
+	}
+
+	printf("Connected to %s:%s\n", ctx->url.host, ctx->url.port);
+	zsock_freeaddrinfo(res);
+	return 0;
+
+error:
+	zsock_close(ctx->sock);
+	ctx->sock = -1;
+	zsock_freeaddrinfo(res);
+	return -errno;
+}
+
+/*
+ * @brief HTTP response callback - processes firmware chunks
+ */
+static int ota_http_response_cb(struct http_response *rsp, enum http_final_call final_data,
+				void *user_data)
+{
+	ARG_UNUSED(final_data);
+	struct app_ctx *ctx = (struct app_ctx *)user_data;
+
+	ctx->http_status_code = rsp->http_status_code;
+	ctx->image_size = rsp->content_range.total;
+
+	if (rsp->body_frag_len > 0) {
+		if (ctx->flash_buffer_len + rsp->body_frag_len > sizeof(ctx->flash_buffer)) {
+			return -ENODATA;
+		}
+		memcpy(ctx->flash_buffer + ctx->flash_buffer_len, rsp->body_frag_start,
+		       rsp->body_frag_len);
+		ctx->flash_buffer_len += rsp->body_frag_len;
+	}
+
+	return 0;
+}
+
+/*
+ * @brief Download firmware chunk from server
+ */
+static int ota_download_firmware_chunk(struct app_ctx *ctx)
+{
+	char range_header[64];
+	const char *headers[] = {range_header, NULL};
+	int ret = 0;
+	struct http_request req = {
+		.method = HTTP_GET,
+		.url = ctx->url.path,
+		.header_fields = headers,
+		.host = ctx->url.host,
+		.protocol = "HTTP/1.1",
+		.response = ota_http_response_cb,
+		.recv_buf = ctx->http_recv_buffer,
+		.recv_buf_len = sizeof(ctx->http_recv_buffer)
+	};
+
+	ctx->flash_buffer_len = 0;
+
+	/* Prepare headers */
+	snprintf(range_header, sizeof(range_header), "Range: bytes=%d-%d\r\n", ctx->range_start,
+		 ctx->range_start + sizeof(ctx->flash_buffer) - 1);
```

Comment:
Can you confirm the HTTP server is fine when we request `range_end` beyond `image_size`?

Reply:
I would be more comfortable with:

    int range_end = ctx->range_start + sizeof(ctx->flash_buffer) - 1;
    if (ctx->range_start /* image_size not valid on first chunk */) {
        range_end = MIN(range_end, ctx->image_size);
    }

However, if you guarantee the current code work, I fine with that.